### PR TITLE
adjust structure of `derp` section of `hp` 

### PR DESCRIPTION
**Pull request by ramfox**
Includes a more or less one-to-one port of `derp.go` to `derp/derp.rs`.

Should probably be refactored to be more "rust"-y, but I'm still trying to understand how pieces fit together.

### CODE REVIEW

File: src/hp/derp/derp.rs [hunk 1]
```diff
@@ -0,0 +1,175 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+// Package derp implements the Designated Encrypted Relay for Packets (DERP)
```
Comment (dignifiedquire) [inline src/hp/derp/derp.rs:4]: please use module doc comments

File: src/hp/derp/derp.rs [hunk 2]
```diff
@@ -0,0 +1,175 @@
+// Copyright (c) Tailscale Inc & AUTHORS
```
Comment (dignifiedquire) [inline src/hp/derp/derp.rs:1]: drope these please, we don’t have it any of the files and instead add a comment what file this is based on

File: src/hp/derp/derp.rs [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+// protocol.
+//
+// DERP routes packets to clients using curve25519 keys as addresses.
+//
+// DERP is used by Tailscale nodes to proxy encrypted WireGuard
+// packets through the Tailscale cloud servers when a direct path
+// cannot be found or opened. DERP is a last resort. Both side
+// between very aggressive NATs, firewalls, no IPv6, etc? Well, DERP.
+
+use std::io::{Read, Write};
```
Comment (dignifiedquire) [inline src/hp/derp/derp.rs:14]: we need async versions of all these

File: src/hp/derp/derp.rs [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+use std::time::Duration;
+
+use anyhow::{bail, Result};
+use byteorder::{BigEndian, ReadBytesExt, WriteBytesExt};
+use bytes::BytesMut;
+
+// MaxPacketSize is the maximum size of a packet sent over DERP.
```
Comment (dignifiedquire) [inline src/hp/derp/derp.rs:21]: please update the comments to remove the go doc requirement of repeating the name and convert them to /// doc comments

File: src/hp/derp/derp.rs [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+///  * server sends FRAME_SERVER_INFO
+///
+///  Steady state:
+///  * server occasionally sends FRAME_KEEP_ALIVE (or FRAME_PING)
+///  * client responds to any FRAME_PING with a FRAME_PONG
+///  * clients sends FRAME_SEND_PACKET
+///  * server then sends FRAME_RECV_PACKET to recipient
+///
+
+const FRAME_SERVER_KEY: FrameType = 0x01; // 8B magic + 32B public key + (0+ bytes future use)
```
Comment (dignifiedquire) [inline src/hp/derp/derp.rs:65]: please make these comments doc comments

File: src/hp/derp/derp_cfg.rs [hunk 6]
```diff
@@ -0,0 +1,62 @@
+// from tailscale/tailcfg/derpmap.go
```
Comment (dignifiedquire) [inline src/hp/derp/derp_cfg.rs:1]: why call this file cfg when the source is map?

File: src/hp/derp/derphttp.rs [hunk 7]
```diff
@@ -0,0 +1,3 @@
+mod derphttp_client;
+
```
Comment (dignifiedquire) [inline src/hp/derp/derphttp.rs:2]: if the module is called derp, just call this derp/http.rs

File: src/hp/derp/derphttp/derphttp_client.rs [hunk 8]
```diff
@@ -0,0 +1,72 @@
+use std::net::SocketAddr;
```
Comment (dignifiedquire) [inline src/hp/derp/derphttp/derphttp_client.rs:1]: please add a ref to the original file

File: src/hp/derp/derp_cfg.rs (hunk 6 above)
Comment (ramfox) [reply src/hp/derp/derp_cfg.rs:1]: good point

File: src/hp/derp/derp.rs [hunk 9]
```diff
@@ -62,114 +59,132 @@ type FrameType = u8;
...
+async fn write_frame_header(
+    mut writer: impl AsyncWrite + Unpin,
     frame_type: FrameType,
     frame_len: u32,
-) -> std::io::Result<()> {
-    writer.write_u8(frame_type)?;
-    writer.write_u32::<BigEndian>(frame_len)
+) -> Result<()> {
+    writer.write(&[frame_type]).await?;
+    writer.write(&frame_len.to_be_bytes()).await?;
```
Comment (dignifiedquire) [inline src/hp/derp/derp.rs:172]: same for writing, see `AsyncWriteExt`

File: src/hp/derp/derp_client.rs [hunk 10]
```diff
@@ -1,3 +1,4 @@
+//! based on tailscale/derp/derp_client.go
```
Comment (dignifiedquire) [inline src/hp/derp/derp_client.rs:1]: should probably be called `derp/client.rs`

File: src/hp/derp/derp.rs [hunk 11]
```diff
@@ -0,0 +1,190 @@
+//! Package derp implements the Designated Encrypted Relay for Packets (DERP)
```
Comment (dignifiedquire) [inline src/hp/derp/derp.rs:1]: this files content should go into `hp/derp.rs`, double derp is a bit much 😅

**Pull request merged**